File: packages/fsm-core-async-op-worker/src/worker-sdk/typescript/sdk.ts
// TypeScript worker SDK: connects to the gateway's sidecar Unix socket,
// registers verified actors, and serves invoke requests.
// Ported from the polygot-lang-ipc-worker prototype's
// worker-sdk/typescript/src/cli.ts, with changes:
//
// - Reuses this package's own sidecar/protocol.ts for framing/envelopes
//   instead of duplicating the wire protocol (the prototype had to
//   duplicate it because the gateway and each worker-sdk were separate
//   deployable units; here they're the same monorepo package).
// - Routes by `actorKey()`'s
//   `parentFsmName@parentFsmVersion@fsmType@fsmName@fsmVersion` instead of a
//   free-form function name, matching the activity contract this gateway
//   speaks.
// - Actor discovery is validate-async-operation.ts's
//   `validateAsyncOperationFromFoldersTypescript` (this repo's real FSM
//   actor convention — see check_fn.ts), not a standalone directory scan.
//   `ActorWorker` takes the verified `ActorPluginValidationResult[]`
//   directly and dynamically imports each `fsmModulePath`/`method` pair
//   itself.
//
// This is the reference implementation for a compiled-language worker SDK
// (e.g. Rust) to follow — see SPEC-001's acceptance criteria.

import { toFileUrl } from "@std/path";
import { getLogger } from "@logtape/logtape";
import type { ActorPluginValidationResult } from "@pgfsm/compiler";
import {
  actorKey,
  type InvokeBody,
  type InvokeErrorBody,
  type InvokeResultBody,
  makeEnvelope,
  readFrame,
  type RegisterAckBody,
  type RegisterBody,
  type RegisteredActor,
  writeFrame,
} from "../../sidecar/protocol.ts";

const logger = getLogger([
  "@pgfsm/worker",
  "async-op-worker-gateway",
  "worker-sdk-ts",
]);

export type ActorHandler = (input: unknown) => unknown | Promise<unknown>;

const DEFAULT_HEARTBEAT_MS = 5_000;

export interface ActorWorkerOptions {
  workerId: string;
  language: string;
  gatewaySocketPath: string;
  heartbeatMs?: number;
}

// Connects to the gateway's sidecar socket, registers `verifiedActors`
// (already filtered to `isVerified === true` by the caller), and serves
// invoke requests until the gateway unregisters this worker or `stop()` is
// called.
export class ActorWorker {
  private conn: Deno.Conn | null = null;
  private stopped = false;
  private readonly handlers = new Map<string, ActorHandler>();

  constructor(
    private readonly options: ActorWorkerOptions,
    private readonly verifiedActors: ActorPluginValidationResult[],
  ) {}

  async run(): Promise<void> {
    if (this.verifiedActors.length === 0) {
      throw new Error("no actors to register, refusing to start worker");
    }

    const registeredActors: RegisteredActor[] = [];
    for (const actor of this.verifiedActors) {
      const mod = await import(toFileUrl(actor.fsmModulePath).href);
      const handler = mod[actor.method];
      if (typeof handler !== "function") {
        throw new Error(
          `'${actor.method}' is not exported as a function from ${actor.fsmModulePath}`,
        );
      }

      this.handlers.set(
        actorKey(
          actor.parentFsmName,
          actor.parentFsmVersion,
          actor.fsmType,
          actor.fsmName,
          actor.fsmVersion,
        ),
        handler as ActorHandler,
      );
      registeredActors.push({
        parentFsmName: actor.parentFsmName,
        parentFsmVersion: actor.parentFsmVersion,
        fsmType: actor.fsmType,
        fsmName: actor.fsmName,
        fsmVersion: actor.fsmVersion,
      });
    }

    const conn = await Deno.connect({
      transport: "unix",
      path: this.options.gatewaySocketPath,
    });
    this.conn = conn;

    const registerBody: RegisterBody = {
      worker_id: this.options.workerId,
      language: this.options.language,
      protocol_version: "1.0",
      actors: registeredActors,
    };

    await writeFrame(
      conn,
      makeEnvelope(
        "register",
        `worker:${this.options.workerId}`,
        "gateway",
        registerBody as unknown as Record<string, unknown>,
      ),
    );

    const ack = await readFrame(conn);
    if (!ack || ack.type !== "register_ack") {
      throw new Error(`expected register_ack but got ${ack?.type ?? "EOF"}`);
    }
    const ackBody = ack.body as unknown as RegisterAckBody;
    if (!ackBody.accepted) {
      throw new Error("gateway rejected registration");
    }

    logger.info(
      "Worker {workerId} registered {count} actor(s) with the gateway",
      { workerId: this.options.workerId, count: registeredActors.length },
    );

    const heartbeat = this.heartbeatLoop(conn);

    try {
      await this.serveLoop(conn);
    } finally {
      this.stopped = true;
      await heartbeat.catch(() => {});
      try {
        conn.close();
      } catch {
        // already closed
      }
    }
  }

  stop(): void {
    this.stopped = true;
    try {
      this.conn?.close();
    } catch {
      // already closed
    }
  }

  private async heartbeatLoop(conn: Deno.Conn): Promise<void> {
    const intervalMs = this.options.heartbeatMs ?? DEFAULT_HEARTBEAT_MS;
    while (!this.stopped) {
      await new Promise((resolve) => setTimeout(resolve, intervalMs));
      if (this.stopped) {
        break;
      }
      await writeFrame(
        conn,
        makeEnvelope(
          "heartbeat",
          `worker:${this.options.workerId}`,
          "gateway",
          { worker_id: this.options.workerId },
        ),
      );
    }
  }

  private async serveLoop(conn: Deno.Conn): Promise<void> {
    while (!this.stopped) {
      const envelope = await readFrame(conn);
      if (!envelope) {
        break;
      }

      if (envelope.type === "cancel") {
        continue;
      }
      if (envelope.type === "unregister") {
        break;
      }
      if (envelope.type !== "invoke") {
        continue;
      }

      await this.handleInvoke(conn, envelope.body as unknown as InvokeBody);
    }
  }

  private async handleInvoke(conn: Deno.Conn, body: InvokeBody): Promise<void> {
    const key = actorKey(
      body.parent_fsm_name,
      body.parent_fsm_version,
      body.fsm_type,
      body.fsm_name,
      body.fsm_version,
    );
    const handler = this.handlers.get(key);
    const started = performance.now();

    if (!handler) {
      await this.sendError(
        conn,
        body.invoke_id,
        "NOT_FOUND",
        `actor not found: ${key}`,
      );
      return;
    }

    try {
      const output = await Promise.resolve(handler(body.input));
      const durationMs = Math.max(0, Math.round(performance.now() - started));
      const result: InvokeResultBody = {
        invoke_id: body.invoke_id,
        output,
        duration_ms: durationMs,
      };
      await writeFrame(
        conn,
        makeEnvelope(
          "invoke_result",
          `worker:${this.options.workerId}`,
          "gateway",
          result as unknown as Record<string, unknown>,
        ),
      );
    } catch (error) {
      await this.sendError(
        conn,
        body.invoke_id,
        "INTERNAL",
        error instanceof Error ? error.message : "unknown worker error",
      );
    }
  }

  private async sendError(
    conn: Deno.Conn,
    invokeId: string,
    code: string,
    message: string,
  ): Promise<void> {
    const errorBody: InvokeErrorBody = {
      invoke_id: invokeId,
      error: { code, message, retriable: false },
    };
    await writeFrame(
      conn,
      makeEnvelope(
        "invoke_error",
        `worker:${this.options.workerId}`,
        "gateway",
        errorBody as unknown as Record<string, unknown>,
      ),
    );
  }
}

File: packages/fsm-core-async-op-worker/src/worker-sdk/python/sdk.py
"""Python worker SDK: connects to the gateway's sidecar Unix socket,
registers verified actors, and serves invoke requests.

Python counterpart of ../typescript/sdk.ts's ActorWorker — same wire
protocol (protocol.py, ported from ../../sidecar/protocol.ts), same
actor_key() identity
(parent_fsm_name@parent_fsm_version@fsm_type@fsm_name@fsm_version), same
register -> heartbeat -> serve lifecycle, and it dynamically loads each
verified actor's module/function itself rather than taking a pre-built
handler map. Uses a background thread for heartbeats (matching the
polygot-lang-ipc-worker prototype's Python worker-sdk) instead of the
TypeScript version's async Deno.Conn loop — plain blocking sockets + a
heartbeat thread is the natural Python shape for the same protocol.

This is the reference implementation alongside worker-sdk/typescript for a
compiled-language worker SDK (e.g. Rust) to follow — see SPEC-001's
acceptance criteria.
"""

from __future__ import annotations

import asyncio
import importlib.util
import socket
import threading
import time
from types import ModuleType
from typing import Any, Callable, Dict, List, Optional

from protocol import actor_key, make_envelope, read_frame, write_frame
from validate_async_operation import ActorPluginValidationResult

ActorHandler = Callable[[Any], Any]

DEFAULT_HEARTBEAT_MS = 5000


class ProtocolError(Exception):
    pass


def _load_module(module_path: str) -> ModuleType:
    spec = importlib.util.spec_from_file_location("_pgfsm_actor", module_path)
    if spec is None or spec.loader is None:
        raise ImportError(f"Cannot load module from {module_path}")
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)  # type: ignore[union-attr]
    return module


class ActorWorker:
    def __init__(
        self,
        worker_id: str,
        gateway_socket_path: str,
        verified_actors: List[ActorPluginValidationResult],
        heartbeat_ms: int = DEFAULT_HEARTBEAT_MS,
    ) -> None:
        self.worker_id = worker_id
        self.language = "python"
        self.gateway_socket_path = gateway_socket_path
        self.verified_actors = verified_actors
        self.heartbeat_ms = heartbeat_ms

        self._sock: Optional[socket.socket] = None
        self._handlers: Dict[str, ActorHandler] = {}
        self._stopped = False

    def run(self) -> None:
        if not self.verified_actors:
            raise ValueError("no actors to register, refusing to start worker")

        registered_actors = []
        for actor in self.verified_actors:
            module = _load_module(actor.fsm_module_path)
            handler = getattr(module, actor.method, None)
            if not callable(handler):
                raise ValueError(
                    f"'{actor.method}' is not exported as a function "
                    f"from {actor.fsm_module_path}"
                )

            key = actor_key(
                actor.parent_fsm_name,
                actor.parent_fsm_version,
                actor.fsm_type,
                actor.fsm_name,
                actor.fsm_version,
                actor.fsm_language,
            )
            self._handlers[key] = handler
            registered_actors.append(
                {
                    "parentFsmName": actor.parent_fsm_name,
                    "parentFsmVersion": actor.parent_fsm_version,
                    "fsmType": actor.fsm_type,
                    "fsmName": actor.fsm_name,
                    "fsmVersion": actor.fsm_version,
                    "fsmLanguage": actor.fsm_language,
                }
            )

        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        sock.connect(self.gateway_socket_path)
        self._sock = sock

        register_body = {
            "worker_id": self.worker_id,
            "language": self.language,
            "protocol_version": "1.0",
            "actors": registered_actors,
        }
        write_frame(
            sock,
            make_envelope(
                "register", f"worker:{self.worker_id}", "gateway", register_body
            ),
        )

        ack = read_frame(sock)
        if ack is None or ack.get("type") != "register_ack":
            raise ProtocolError(
                f"expected register_ack but got {ack.get('type') if ack else 'EOF'}"
            )
        if not ack.get("body", {}).get("accepted", False):
            raise ProtocolError("gateway rejected registration")

        print(
            f"Worker {self.worker_id} registered {len(registered_actors)} actor(s) "
            "with the gateway"
        )

        heartbeat_thread = threading.Thread(target=self._heartbeat_loop, daemon=True)
        heartbeat_thread.start()

        try:
            self._serve_loop()
        finally:
            self._stopped = True
            try:
                sock.close()
            except OSError:
                pass

    def stop(self) -> None:
        self._stopped = True
        if self._sock is not None:
            try:
                self._sock.close()
            except OSError:
                pass

    def _heartbeat_loop(self) -> None:
        assert self._sock is not None
        while not self._stopped:
            time.sleep(self.heartbeat_ms / 1000)
            if self._stopped:
                break
            try:
                write_frame(
                    self._sock,
                    make_envelope(
                        "heartbeat",
                        f"worker:{self.worker_id}",
                        "gateway",
                        {"worker_id": self.worker_id},
                    ),
                )
            except OSError:
                self._stopped = True
                break

    def _serve_loop(self) -> None:
        assert self._sock is not None
        while not self._stopped:
            envelope = read_frame(self._sock)
            if envelope is None:
                break

            msg_type = envelope.get("type")
            if msg_type == "cancel":
                continue
            if msg_type == "unregister":
                break
            if msg_type != "invoke":
                continue

            self._handle_invoke(envelope.get("body", {}))

    def _handle_invoke(self, body: Dict[str, Any]) -> None:
        invoke_id = body.get("invoke_id")
        key = actor_key(
            body.get("parent_fsm_name", ""),
            body.get("parent_fsm_version", ""),
            body.get("fsm_type", ""),
            body.get("fsm_name", ""),
            body.get("fsm_version", ""),
            body.get("fsm_language", ""),
        )
        handler = self._handlers.get(key)

        if handler is None:
            self._send_error(invoke_id, "NOT_FOUND", f"actor not found: {key}")
            return

        started = time.perf_counter()
        try:
            if asyncio.iscoroutinefunction(handler):
                output = asyncio.run(handler(body.get("input")))
            else:
                output = handler(body.get("input"))
            duration_ms = max(0, round((time.perf_counter() - started) * 1000))
            self._send(
                make_envelope(
                    "invoke_result",
                    f"worker:{self.worker_id}",
                    "gateway",
                    {
                        "invoke_id": invoke_id,
                        "output": output,
                        "duration_ms": duration_ms,
                    },
                )
            )
        except Exception as exc:  # noqa: BLE001 — reported to the gateway, not raised
            self._send_error(invoke_id, "INTERNAL", str(exc))

    def _send_error(self, invoke_id: Optional[str], code: str, message: str) -> None:
        self._send(
            make_envelope(
                "invoke_error",
                f"worker:{self.worker_id}",
                "gateway",
                {
                    "invoke_id": invoke_id,
                    "error": {"code": code, "message": message, "retriable": False},
                },
            )
        )

    def _send(self, envelope: Dict[str, Any]) -> None:
        assert self._sock is not None
        write_frame(self._sock, envelope)
